Middleware datasources: topology graph

Jiri-Kremser opened the pull request:
fixes issue #8628 

![screenshot from 2016-05-11 19-38-06](https://cloud.githubusercontent.com/assets/535866/15190403/fe8f1a30-17af-11e6-96c6-ab03878b531a.png)

Most of the changed lines is due to re-recording the JSON that is returned from the server. The data was stale, because the commit f48dce286d9d0080a0d63475532cdddd0b0c8857 that adds the tests for middleware topology graph wasn't reflecting the commit c063ed1fea40e8486b90fd26dd879fc9572a3caa that changed the JSON data (the data was recorded before the commit). Hopefully, it makes sense.

@miq-bot add_label topology, providers/hawkular, wip
@miq-bot remove_label gem changes

abonas commented:
I know this is still wip, but please make the commit message more clear, e.g. "add middleware datasource entity to the topology graph" or something similar.

abonas commented:
@ammendonca please sync with @Jiri-Kremser on this PR since it is also doing changes around icons and topology.

abonas commented on app/models/mixins/ui_service_mixin.rb:
```diff
@@ -1,21 +1,24 @@
 module UiServiceMixin
+  # rubocop:disable Metrics/LineLength
```
I'm not in favor of adding rubocop related comments inside the source code

Jiri-Kremser replied:
ok, removed. We will have some Rubocop offenses in return :]

abonas replied:
yes I agree it's not ideal, however mixing source code and external validations that are ran on it is also not a good practice. today it's rubocop, tomorrow it's something else, putting those comments in code will force doing a major cleanup if rubocop is replaced. if there are some things and rules that need to be fine tuned in rubocop, that can and should be discussed.

abonas commented on app/services/middleware_topology_service.rb:
```diff
@@ -33,14 +36,14 @@ def entity_display_type(entity)
-    data[:icon] = entity.decorate.try(:listicon_image)
+    data[:status] = 'Unknown'
+    data[:display_kind] = entity_display_type(entity)
+    data[:icon] = entity.decorate.try(:listicon_image) unless entity.kind_of?(MiddlewareDatasource)
```
perhaps worth here to do it slightly differently: try to get the font icon from decorator, if empty - get the image from decorator. that can also solve adding things to ui mixin, because currently the icon is defined twice - once in the decorator and once in uimixin

Jiri-Kremser replied:
Although I like your proposal, it's not easily doable, the topology graph is an svg image and it doesn't support the CSS pseudo elements (:before :after) https://stackoverflow.com/questions/14938201/unexpected-duplication-of-after-pseudo-element-assigned-to-svgtext-working-wi

That's actually why the glyphs are stored in the code as unicode characters and not as thereadable `fa fa-something` css class "selectors"

btw. I rewrote it to this, but then figured out, that it can work only if the decorator will return the unicode char directly and not the `fa fa-database`, but this will break the other occurences that depend on this form. 
```ruby
    # fonticon takes precedence before the listicon_image
    data[:icon] = if icon = entity.decorate.try(:fonticon)
                    {
                      :type  => :glyph,
                      :value => icon
                    }
                  elsif icon = entity.decorate.try(:listicon_image)
                    {
                      :type  => :image,
                      :value => "/assets/svg/#{icon}.svg"
                    }
                  end
    data
```

I'd stick with the `UiServiceMixin`, 1 global or split it into smaller ones. Both approaches have it's own benefits. IMO I am spending too much time on this :)

Jiri-Kremser replied:
I've done it in a way that the fonticon from the decorator is honored, but the UTF char is taken from the `UiServiceMixin`, there are new helper methods on this module for the mapping

abonas commented on spec/javascripts/controllers/middleware/middleware_topology_controller_spec.js:
```diff
@@ -65,28 +73,42 @@ describe('middlewareTopologyController', function () {
       expect($controller.class_name(mw_manager)).toContain("vendor-hawkular");
       expect($controller.class_name(mw_server)).toContain("vendor-hawkular");
       expect($controller.class_name(mw_deployment)).toContain("middleware_deployment_war");
+      expect($controller.class_name(mw_datasource)).toBeUndefined()
```
why undefined?

Jiri-Kremser replied:
because https://github.com/ManageIQ/manageiq/pull/8630/files#diff-5e5229c950a2f71e17950e9591975c70L96

hmm, not very intuitive, I can add a comment, or change the method.. perhaps some better name..

abonas replied:
I am following the link to the code diff, but it looks like it's bringing me to an unrelated place, or I don't understand the relation to the code in the test above.

Jiri-Kremser replied:
`class_name()`:

```javascript
      self.class_name = function class_name(d) {
          var class_name = d.item.icon;
          return class_name;
      }
```

For datasources the icon is not set, because it's done via the `foo.kind_of?(MiddlewareDatasource)` which is not perfect, I'll change that to use the font icon from the decorator.

Jiri-Kremser commented:
@miq-bot remove_label wip

abonas commented on app/assets/javascripts/controllers/middleware_topology/middleware_topology_controller.js:
```diff
@@ -37,12 +38,12 @@ function MiddlewareTopologyCtrl($scope, $http, $interval, $location, topologySer
 
-       if($scope.checkboxModel.value) {
-            vertices.selectAll("text")
+       if ($scope.checkboxModel.value) {
```
How is this change related to adding data source support to topology? if there's another issue that needs fixing here, it is worth to submit it as a separate PR.

Jiri-Kremser replied:
font awesome glyphs need to be added as text

abonas replied:
:+1:

abonas commented on app/assets/javascripts/controllers/middleware_topology/middleware_topology_controller.js:
```diff
@@ -112,8 +150,25 @@ function MiddlewareTopologyCtrl($scope, $http, $interval, $location, topologySer
+    }
 
-    };
+    self.getTextPosition = function getTextPosition(d) {
```
It seems like this is an enhancement to place better the labels on the graph - which is great, but not related to adding datasources to topology. I suggest submitting a separate PR with this. worth to add the defaults in https://github.com/ManageIQ/manageiq/blob/master/app/assets/javascripts/services/topology_service.js
And then it can influence all the topologies if needed, not only middleware topology.

Jiri-Kremser replied:
Hmm, in fact it's used only for positioning the glyphs, it's not related to the labels (but I'll double check). I'll move that to the css and remove this function. Good catch

abonas commented on app/assets/javascripts/controllers/middleware_topology/middleware_topology_controller.js:
```diff
@@ -112,8 +150,25 @@ function MiddlewareTopologyCtrl($scope, $http, $interval, $location, topologySer
+
+    self.getGlyphFeatures = function getGlyphFeatures(d) {
+        switch (d.item.kind) {
+            case "MiddlewareDatasource" :
```
this is better to be placed in a css file and not in code. worth to add a new css file for mw topology.
Please see examples including how glyph style is handled [here](https://github.com/ManageIQ/manageiq/blob/master/app/assets/stylesheets/container_topology.css ) and [here](https://github.com/ManageIQ/manageiq/blob/master/app/assets/stylesheets/topology.css)

Jiri-Kremser replied:
Yeah, this makes sense. I was considering it, but then went with the faster solution.

abonas replied:
:+1:

abonas commented on app/models/mixins/ui_service_mixin.rb:
```diff
@@ -1,24 +1,44 @@
+      :Host                 => {:type => "glyph", :icon => "\uE600", :fontfamily => "PatternFlyIcons-webfont"}, # pficon-screen
+      :Vm                   => {:type => "glyph", :icon => "\uE90f", :fontfamily => "PatternFlyIcons-webfont"}, # pficon-virtual-machine
+      :MiddlewareDatasource => {:type => "glyph", :icon => "\uF1C0", :fontfamily => "FontAwesome",
+                                :css_class => "fa fa-database"},
```
why is the css_class needed - what is different from other glyphs here in the file?

Jiri-Kremser replied:
other glyphs use the approach I had in a previous version of my PR where you commented that the glyph should be taken from the decorator. However, in the decorator, it's stored in the `fa fa-database` form, hence this occurrence. It basically provides the mapping `fa fa-database` -> utf char

The reason for this was already described here https://github.com/ManageIQ/manageiq/pull/8630#discussion_r63847403

abonas commented on app/services/middleware_topology_service.rb:
```diff
@@ -33,14 +36,28 @@ def entity_display_type(entity)
+
+    # fonticon takes precedence before the listicon_image
+    # and both methods on decorator takes precedence before the global icons hash from the UiServiceMixin
+    data[:icon] = if !(icon = entity.decorate.try(:fonticon)).nil?
```
why is the icon added here to every entity if it's already added once for each entity type in the icons variable that is passed to frontend?

Jiri-Kremser replied:
I think, it's obvious from the comment. If the entity has a decorator and the decorator's `fonticon` method doesn't return `nil`, it's used, then the fallback is `listicon_image` and then it falls back to the entity type based approach using the `icons` (https://git.io/vrapO)

abonas replied:
You are correct. I am starting to think that putting font icons in decorators is not a good idea: 
it doesn't contain font family there, it has no utf char there that is needed here, etc.
and there is again duplication - decorator and uimixin both have the same information that needs to be maintained twice. plus, the icons are still repeated in every entity in topology json, hence increasing the json size.
@pilhuhn I'm adding you to this discussion, I sense we need some "cleaner" approach for font icons and would like to hear your ideas.

pilhuhn commented on app/assets/javascripts/controllers/middleware_topology/middleware_topology_controller.js:
```diff
@@ -71,17 +72,51 @@ function MiddlewareTopologyCtrl($scope, $http, $interval, $location, topologySer
+        added.append("text")
+            .each(function(d) {
+                var iconInfo = self.getIcon(d);
+                if (iconInfo.type != 'glyph') return;
```
Please put the `return` on the next line

Jiri-Kremser replied:
fixed

abonas commented on app/assets/stylesheets/middleware_topology.css:
```diff
@@ -0,0 +1,4 @@
+.middleware .kube-topology g.MiddlewareDatasource text.glyph {
```
you need to add this file dependency to https://github.com/ManageIQ/manageiq/blob/master/app/assets/stylesheets/application.css

abonas commented on app/assets/javascripts/controllers/middleware_topology/middleware_topology_controller.js:
```diff
@@ -71,17 +72,51 @@ function MiddlewareTopologyCtrl($scope, $http, $interval, $location, topologySer
+                return d.item.name;
+            })
+            .attr('class', 'attached-label')
+            .style("font-size", function(d) {
```
is this needed given the following css?
https://github.com/ManageIQ/manageiq/blob/master/app/assets/stylesheets/topology.css#L47

Jiri-Kremser replied:
fixed

pilhuhn commented:
I don't like that UIMixin - I think those definitions should go into the decorators. The topology services should then probably ask those icon infos and cache in a hash<entity-type,hash<kind,value>> where kind is something like png, font, glyph etc.

We may also want to talk with the UI team to potentially revisit the use of font-icons wrt svg (there was a good article on why svg is better anyway: https://github.com/blog/2112-delivering-octicons-with-svg ).

abonas commented:
> I don't like that UIMixin - I think those definitions should go into the decorators. The topology services should then probably ask those icon infos and cache in a hash> where kind is something like png, font, glyph etc.
> 
> We may also want to talk with the UI team to potentially revisit the use of font-icons wrt svg (there was a good article on why svg is better anyway: https://github.com/blog/2112-delivering-octicons-with-svg ).

@pilhuhn 
The UIMixin was added before there was an advice to use decorators.
In any case, the text icons are used because at least for containers, they come come from pf (except for the providers icon for kubernetes, atomic and openshift)
https://www.patternfly.org/styles/icons/
And using the decorator still means we need there svg, text icon, its utf representation and its font family. pls see here https://github.com/ManageIQ/manageiq/issues/8780#issuecomment-221513200

abonas commented:
@himdel @martinpovolny I'd like to hear your opinion to the discussion regarding decorators, mixins and font icons. I believe there should be a good single approach used by everyone for this. please see above https://github.com/ManageIQ/manageiq/pull/8630#issuecomment-221563916

martinpovolny commented:
Ui mixin is better than the previous approach. It gives a way to put the information to a single place.

Decorators are a way that we are trying to use to address e.g. provider plugability w/o putting UI stuff into the backend. We have introduced them recently to several places in the UI.

As long as there's only 1 such method I am ok with the UI mixin approach or a decorator. Once there's 2 or more there should be a decorator for sure.

pilhuhn commented:
@martinpovolny One issue with *this* UiMixin is that it is one mixin for 2 providers, which creates another coupling and is against the idea of pluggable providers.

abonas commented:
@Jiri-Kremser, please check what is the status of this PR and how it can proceed towards completion and merging. Does it require a further technical discussion (please ping relevant people to finalize it), rebase, etc. thank you.

Jiri-Kremser commented:
@abonas I've rebased and fixed a missing icon. imho, this is ready to be merged. I checked, the datasource icon is not missing in the topology graph. Other icons are missing in the detail views (not in the topology), but it's not related to this PR

Jiri-Kremser commented:
@miq-bot remove_label gem changes

abonas commented on app/assets/javascripts/controllers/middleware_topology/middleware_topology_controller.js:
```diff
@@ -10,7 +10,7 @@ function MiddlewareTopologyCtrl($scope, $http, $interval, $location, topologySer
     $scope.refresh = function() {
         var id;
-        if ($location.absUrl().match("show/$") || $location.absUrl().match("show$")) {
+        if ($location.absUrl().match('show/$') || $location.absUrl().match('show$')) {
```
those kind of changes make the review harder as they are not related to the PR itself. please try to separate those fixes in the future to another PR.

Jiri-Kremser replied:
I didn't want to end up in a state that new code will use the strings with single quotes and the old one the double quotes, so I did the auto replace. Sorry, that it makes the review more difficult, but for me the consistency is simply more.

btw.
https://github.com/ManageIQ/manageiq/issues/8781#issuecomment-221531748
https://github.com/hawkular/hawkular-client-ruby/pull/77#discussion_r67828729

abonas commented on app/assets/javascripts/controllers/middleware_topology/middleware_topology_controller.js:
```diff
@@ -22,6 +22,7 @@ function MiddlewareTopologyCtrl($scope, $http, $interval, $location, topologySer
             $scope.items = data.data.items;
             $scope.relations = data.data.relations;
             $scope.kinds = data.data.kinds;
+            $scope.icons = data.data.icons;
```
why are the icons places on scope? does Angular need them? aren't they used internally here in the controller?

Jiri-Kremser replied:
For being able to test it properly, iirc, the `getIcon` didn't work properly in this context w/o it: https://github.com/ManageIQ/manageiq/pull/8630/files#diff-73ef271f9f62c257d152ab3fa2b11221R109

abonas replied:
have you checked containers topology impl? iirc I didn't use scope there and was able to test the icons

Jiri-Kremser replied:
I'll check

abonas commented on app/assets/javascripts/controllers/middleware_topology/middleware_topology_controller.js:
```diff
@@ -54,66 +55,100 @@ function MiddlewareTopologyCtrl($scope, $http, $interval, $location, topologySer
+                }
+                $(this).text(iconInfo.icon)
+                    .attr('class', 'glyph')
+                    .attr('style', 'font-family:' +iconInfo.fontfamily+';')
```
missing space after the + sign

abonas commented on app/assets/javascripts/controllers/middleware_topology/middleware_topology_controller.js:
```diff
@@ -54,66 +55,100 @@ function MiddlewareTopologyCtrl($scope, $http, $interval, $location, topologySer
+            });
+        // possible glyphs
+        added.append('text')
+            .each(function(d) {
```
I think everything here should be 2 spaces (tab) and not 4 spaces

Jiri-Kremser replied:
Right, I know about it, however, the whole file uses the 4 spaces. I had it actually formatted locally, but then I went with 4 spaces not the change the whole file. I am +1 on doing this and I can do that if you want, but this will make the review harder.

abonas commented on app/assets/stylesheets/middleware_topology.css:
```diff
@@ -0,0 +1,9 @@
+    fill: #81B0C8;
+}
+
+.middleware .kube-topology g text.attached-label {
```
isn't this style already in topology.css? (the general one)

Jiri-Kremser replied:
looking, it's done differently than in the `container_topology_controller.js`, perhaps the `show_hide_names` method can be factored out to some "common JS something". For now, I will do the same as controller for containers does -> adding/removing the visible class and use the common css instead of adding style `display: block/none`

If I do that, then yes, this style can go away, otherwise it's needed, because w/o it the font size is 18px

abonas replied:
I'm definitely in favor of doing a common "display names" thingy for all topologies in separate PR.

abonas commented on app/assets/javascripts/controllers/middleware_topology/middleware_topology_controller.js:
```diff
@@ -54,66 +56,100 @@ function MiddlewareTopologyCtrl($scope, $http, $interval, $location, topologySer
-
-        added.selectAll("title").text(function(d) {
-            return topologyService.tooltip(d).join("\n");
+            .attr('y', function(d) { return self.getCircleDimensions(d).y})
```
the dimensions (x,y, height, width) are set even if the image is null? (in glyph use case)

Jiri-Kremser replied:
Right, this can be fixed by the D3 `each()` function similarly as in here https://git.io/voPqb The code will be more complex though, but the resulting HTML will be better => potential speedup. So it's worth doing

abonas commented:
LGTM. 
@martinpovolny @himdel please have a look, in particular on the js controller changes. thanks!

abonas commented:
@miq-bot assign @chessbyte

himdel commented on app/assets/javascripts/controllers/middleware_topology/middleware_topology_controller.js:
```diff
@@ -54,66 +55,96 @@ function MiddlewareTopologyCtrl($scope, $http, $interval, $location, topologySer
+                    return;
+                }
+                $(this).attr('href', iconInfo.icon)
+                .attr('y', self.getCircleDimensions(d).y)
```
broken formatting here..

And the whole file should be indentet 2 spaces, not 4..

abonas replied:
@himdel have you seen [this comment](https://github.com/ManageIQ/manageiq/pull/8630#discussion_r68060608)? seems like it is ok to keep it as it is for now

himdel replied:
I have now :) but there's still the issue of those `.attr` lines being indented wrong..

Jiri-Kremser replied:
Fixed, I'll do the spacing in a next PR. I'd love to rewrite it to the ES6 (not sure if it's possible now)

so that all those:
```javascript
  .attr('r', function(d) {
    return self.getCircleDimensions(d).r;
  })
```

would become:
```javascript
  .attr('r', (d) => self.getCircleDimensions(d).r)
```

himdel replied:
@Jiri-Kremser thanks :)

As for ES6, it is possible.. with the limitation that it will go through babel5 and the file needs to be `*.es6` not `*.js`. (And the specs do need to be ES5 for now.)

We should hopefully get the tooling up to date before the Euwe release..

Jiri-Kremser commented:
thanks for merge
